// @zenstackhq/sdk/code-gen.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.emitProject = exports.saveProject = exports.createProject = void 0;
const prettier_1 = __importDefault(require("prettier"));
const ts_morph_1 = require("ts-morph");
const types_1 = require("./types");
const formatOptions = {
    trailingComma: 'all',
    tabWidth: 4,
    printWidth: 120,
    bracketSpacing: true,
    semi: true,
    singleQuote: true,
    useTabs: false,
    parser: 'typescript',
};
function formatFile(sourceFile) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const content = sourceFile.getFullText();
            const formatted = yield prettier_1.default.format(content, formatOptions);
            sourceFile.replaceWithText(formatted);
            yield sourceFile.save();
        }
        catch (_a) {
            /* empty */
        }
    });
}
/**
 * Creates a TS code generation project
 */
function createProject(options) {
    return new ts_morph_1.Project({
        compilerOptions: Object.assign({ target: ts_morph_1.ScriptTarget.ES2016, module: ts_morph_1.ModuleKind.CommonJS, esModuleInterop: true, declaration: true, strict: true, skipLibCheck: true, noEmitOnError: true }, options),
    });
}
exports.createProject = createProject;
/**
 * Persists a TS project to disk.
 */
function saveProject(project) {
    return __awaiter(this, void 0, void 0, function* () {
        yield Promise.all(project.getSourceFiles().map((sf) => __awaiter(this, void 0, void 0, function* () {
            yield formatFile(sf);
        })));
        yield project.save();
    });
}
exports.saveProject = saveProject;
/**
 * Emit a TS project to JS files.
 */
function emitProject(project) {
    return __awaiter(this, void 0, void 0, function* () {
        const errors = project.getPreEmitDiagnostics().filter((d) => d.getCategory() === ts_morph_1.DiagnosticCategory.Error);
        if (errors.length > 0) {
            console.error('Error compiling generated code:');
            console.error(project.formatDiagnosticsWithColorAndContext(errors.slice(0, 10)));
            yield project.save();
            throw new types_1.PluginError('', `Error compiling generated code`);
        }
        const result = yield project.emit();
        const emitErrors = result.getDiagnostics().filter((d) => d.getCategory() === ts_morph_1.DiagnosticCategory.Error);
        if (emitErrors.length > 0) {
            console.error('Some generated code is not emitted:');
            console.error(project.formatDiagnosticsWithColorAndContext(emitErrors.slice(0, 10)));
            yield project.save();
            throw new types_1.PluginError('', `Error emitting generated code`);
        }
    });
}
exports.emitProject = emitProject;
