// @prisma/internals-v5/dist/chunk-63KFPQEY.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var chunk_63KFPQEY_exports = {};
__export(chunk_63KFPQEY_exports, {
  theme: () => theme
});
module.exports = __toCommonJS(chunk_63KFPQEY_exports);
var import_chunk_DJGRU4RH = require("./chunk-DJGRU4RH.js");
var theme = {
  keyword: import_chunk_DJGRU4RH.cyan,
  entity: import_chunk_DJGRU4RH.cyan,
  value: (s) => (0, import_chunk_DJGRU4RH.bold)((0, import_chunk_DJGRU4RH.blue)(s)),
  punctuation: import_chunk_DJGRU4RH.blue,
  directive: import_chunk_DJGRU4RH.cyan,
  function: import_chunk_DJGRU4RH.cyan,
  variable: (s) => (0, import_chunk_DJGRU4RH.bold)((0, import_chunk_DJGRU4RH.blue)(s)),
  string: (s) => (0, import_chunk_DJGRU4RH.bold)((0, import_chunk_DJGRU4RH.green)(s)),
  boolean: import_chunk_DJGRU4RH.yellow,
  number: import_chunk_DJGRU4RH.cyan,
  comment: import_chunk_DJGRU4RH.gray
};
