// @prisma/internals-v5/dist/cli/getSchema.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var getSchema_exports = {};
__export(getSchema_exports, {
  getPrismaConfigFromPackageJson: () => import_chunk_LXMH457V.getPrismaConfigFromPackageJson,
  getRelativeSchemaPath: () => import_chunk_LXMH457V.getRelativeSchemaPath,
  getSchema: () => import_chunk_LXMH457V.getSchema,
  getSchemaDir: () => import_chunk_LXMH457V.getSchemaDir,
  getSchemaPath: () => import_chunk_LXMH457V.getSchemaPath,
  getSchemaPathFromPackageJson: () => import_chunk_LXMH457V.getSchemaPathFromPackageJson,
  getSchemaPathFromPackageJsonSync: () => import_chunk_LXMH457V.getSchemaPathFromPackageJsonSync,
  getSchemaPathInternal: () => import_chunk_LXMH457V.getSchemaPathInternal,
  getSchemaPathSync: () => import_chunk_LXMH457V.getSchemaPathSync,
  getSchemaPathSyncInternal: () => import_chunk_LXMH457V.getSchemaPathSyncInternal
});
module.exports = __toCommonJS(getSchema_exports);
var import_chunk_LXMH457V = require("../chunk-LXMH457V.js");
var import_chunk_J2RYL4JN = require("../chunk-J2RYL4JN.js");
var import_chunk_KXIPL557 = require("../chunk-KXIPL557.js");
var import_chunk_HOUI6XEN = require("../chunk-HOUI6XEN.js");
var import_chunk_RXZEDVTJ = require("../chunk-RXZEDVTJ.js");
var import_chunk_DJGRU4RH = require("../chunk-DJGRU4RH.js");
var import_chunk_FVJ3R4NJ = require("../chunk-FVJ3R4NJ.js");
