// @prisma/internals-v5/dist/utils/fs-functional.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var fs_functional_exports = {};
__export(fs_functional_exports, {
  createDirIfNotExists: () => import_chunk_2IEY3QNF.createDirIfNotExists,
  getFilesInDir: () => import_chunk_2IEY3QNF.getFilesInDir,
  getNestedFoldersInDir: () => import_chunk_2IEY3QNF.getNestedFoldersInDir,
  removeDir: () => import_chunk_2IEY3QNF.removeDir,
  removeEmptyDirs: () => import_chunk_2IEY3QNF.removeEmptyDirs,
  removeFile: () => import_chunk_2IEY3QNF.removeFile,
  writeFile: () => import_chunk_2IEY3QNF.writeFile
});
module.exports = __toCommonJS(fs_functional_exports);
var import_chunk_2IEY3QNF = require("../chunk-2IEY3QNF.js");
var import_chunk_K43CTUPT = require("../chunk-K43CTUPT.js");
var import_chunk_J77ZIDXB = require("../chunk-J77ZIDXB.js");
var import_chunk_4M6RGZMG = require("../chunk-4M6RGZMG.js");
var import_chunk_PN7UG7JB = require("../chunk-PN7UG7JB.js");
var import_chunk_FVJ3R4NJ = require("../chunk-FVJ3R4NJ.js");
